\documentclass[a4paper,parskip=half,10pt,bibtotoc,abstracton,oneside,noindent,DIV15]{scrartcl}
\usepackage{url,textcomp}
\usepackage{listings}
\usepackage{pslatex,color,hyperref,graphicx,titlesec}
\usepackage[OT1]{fontenc}

\titlespacing{\section}{0pt}{*2}{*0.5}

\setlength{\parindent}{0pt}

\title{SeqPig Manual}

\definecolor{lightlightgray}{gray}{0.9}
\definecolor{OliveGreen}{cmyk}{0.64,0,0.95,0.40}

%\usepackage{tex4ht}
\newcommand{\Css}[1]{}

\lstset{
language=C,                             % Code langugage
basicstyle=\ttfamily,                   % Code font, Examples: \footnotesize, \ttfamily
keywordstyle=\color{OliveGreen},        % Keywords font ('*' = uppercase)
commentstyle=\color{gray},              % Comments font
%numbers=left,                           % Line nums position
%numberstyle=\tiny,                      % Line-numbers fonts
%stepnumber=1,                           % Step between two line-numbers
%numbersep=5pt,                          % How far are line-numbers from code
backgroundcolor=\color{lightlightgray}, % Choose background color
frame=none,                             % A frame around the code
tabsize=2,                              % Default tab size
captionpos=b,                           % Caption-position = bottom
breaklines=true,                        % Automatic line breaking?
breakatwhitespace=false,                % Automatic breaks only at whitespace?
showspaces=false,                       % Dont make spaces visible
showtabs=false,                         % Dont make tabls visible
columns=flexible,                       % Column format
morekeywords={__global__, __device__},  % CUDA specific keywords
upquote=true,
}

\lstset{breaklines=true}

\Css{div.lstlisting{font-family: monospace;
    white-space: nowrap; margin-top:0.5em;
    margin-bottom:0.5em;
    color: blue;}}

\Css{ol li { padding: 0; margin: 0.5em; }}

\begin{document}

\maketitle

\tableofcontents
\newpage

\section{Introduction}

SeqPig is a library for Apache Pig \url{http://pig.apache.org/} for the
distributed analysis of large sequencing datasets on Hadoop clusters.  With
SeqPig one can easily take advantage of the advanced high-level features of Pig
to manipulate and analyze sequencing data, thus writing simple scripts that
automatically run as scalable distributed programs on potentially very large
Hadoop clusters.

SeqPig provides a number of features and functionalities.  It provides import
and export functions for file formats commonly used in bioinformatics, as well
as a collection of Pig user-defined-functions (UDF's) that add to Pig
functionality specifically designed for processing aligned and unaligned
sequencing data. Currently SeqPig supports BAM, SAM, FastQ and Qseq input and
output, thanks in part to the functionality provided by the Hadoop-BAM library.

This document can also be found under
\url{http://seqpig.sourceforge.net/}.\\ Contact information:
\url{mailto:seqpig-users@lists.sourceforge.net}

Releases of SeqPig come bundled with Picard/Samtools, which is
developed at the Wellcome Trust Sanger Institute, and Seal, which
is developed
at CRS4. See\\
\url{http://samtools.sourceforge.net/} and 
\url{http://biodoop-seal.sourceforge.net/}
for more details.

\input{seqpig_installation}

\input{seqpig_examples}

\end{document}
